### PR TITLE
`Paywalls`: new `presentationMode` parameter

### PR DESCRIPTION
**Pull request by Lascorbe**
### TLDR
- Added a way to present `PaywallView` either as sheet or full screen
- No breaking changes have been made

### Motivation
I wanted to easily present the PaywallView on full screen.

### Description
I added a new enum `PaywallPresentationMode` which currently has 2 options: `.sheet` and `.fullScreen `. I added this as a parameter to the current public `presentPaywallIfNeeded` modifiers. The default is `sheet` so there are no breaking changes with the current public API.

| Sheet             | Full Screen    |
| ------------- | ------------- |
| Simulator Screenshot - iPhone 15 Pro - 2024-02-05 at 18 08 51  |  Simulator Screenshot - iPhone 15 Pro - 2024-02-05 at 18 44 15 |

PD: I looked for info on how to contribute but it's unclear if I have to open an issue first or not (contributing says open issue, then when opening an issue it says please don't open but contact us at https://app.revenuecat.com/settings/support). I hope this is valid, otherwise let me know how I can proceed.

### CODE REVIEW

File: RevenueCatUI/View+PresentPaywall.swift [hunk 1]
```diff
@@ -16,6 +16,15 @@ import SwiftUI
 
 #if !os(macOS) && !os(tvOS)
 
+/// Presentation options to use with the View modifiers ``presentPaywallIfNeeded``.
+///
+/// ### Related Articles
+/// [Documentation](https://rev.cat/paywalls)
+public enum PaywallPresentationMode {
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:23]: 👍🏻

File: RevenueCatUI/View+PresentPaywall.swift [hunk 2]
```diff
@@ -16,6 +16,15 @@ import SwiftUI
 
 #if !os(macOS) && !os(tvOS)
 
+/// Presentation options to use with the View modifiers ``presentPaywallIfNeeded``.
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:19]: The `presentPaywallIfNeeded` reference won't be found like that. You'd have to do `View.presentPaywallIfNeeded` but even that probably won't work, so let's change it to
```suggestion
/// Presentation options to use with the View modifiers `.presentPaywallIfNeeded`.
```

File: RevenueCatUI/View+PresentPaywall.swift [hunk 3]
```diff
@@ -38,13 +47,15 @@ extension View {
...
     /// - Parameter fonts: An optional `PaywallFontProvider`.
+    /// - Parameter presentationMode: The desired presentation mode of the ``PaywallView``. Defaults to ``.sheet``.
     ///
     /// ### Related Articles
     /// [Documentation](https://rev.cat/paywalls)
     public func presentPaywallIfNeeded(
         requiredEntitlementIdentifier: String,
         offering: Offering? = nil,
         fonts: PaywallFontProvider = DefaultPaywallFontProvider(),
+        presentationMode: PaywallPresentationMode = .sheet,
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:58]: Let's add this extension so that we don't repeat the default everywhere:
```swift
internal extension PaywallPresentationMode {

    static let `default`: Self = .sheet

}
```

File: RevenueCatUI/View+PresentPaywall.swift [hunk 4]
```diff
@@ -16,6 +16,15 @@ import SwiftUI
...
 #if !os(macOS) && !os(tvOS)
 
+/// Presentation options to use with the View modifiers ``presentPaywallIfNeeded``.
+///
+/// ### Related Articles
+/// [Documentation](https://rev.cat/paywalls)
+public enum PaywallPresentationMode {
+    case sheet
+    case fullScreen
+}
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:26]: Nit: our Swift style guide requires spaces around types:
```suggestion

    case sheet
    case fullScreen
    
}
```

File: RevenueCatUI/View+PresentPaywall.swift [hunk 5]
```diff
@@ -233,57 +253,71 @@ private struct PresentingPaywallModifier: ViewModifier {
...
-                    self.restoreCompleted?(customerInfo)
-
-                    if !self.shouldDisplay(customerInfo) {
-                        self.close()
+        Group {
+            switch presentationMode {
+            case .sheet:
+                content
+                    .sheet(item: self.$data, onDismiss: self.onDismiss) { data in
+                        paywallView(data: data)
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:263]: Explicit `self` per our style guide:
```suggestion
                        self.paywallView(data: data)
```

File: RevenueCatUI/View+PresentPaywall.swift [hunk 6]
```diff
@@ -233,57 +253,71 @@ private struct PresentingPaywallModifier: ViewModifier {
...
+                    }
+            case .fullScreen:
+                content
+                    .fullScreenCover(item: self.$data, onDismiss: self.onDismiss) { data in
+                        paywallView(data: data)
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:268]:
```suggestion
                        self.paywallView(data: data)
```

File: RevenueCatUI/View+PresentPaywall.swift [hunk 7]
```diff
@@ -233,57 +253,71 @@ private struct PresentingPaywallModifier: ViewModifier {
...
+                Logger.debug(Strings.displaying_paywall)
+                
+                self.data = .init(customerInfo: info)
+            } else {
+                Logger.debug(Strings.not_displaying_paywall)
+            }
+        }
+    }
+    
+    private func paywallView(data: Data) -> some View {
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:287]: Probably simpler:
```suggestion
    private func paywallView(_ data: Data) -> some View {
```

**Review by NachoSoto — COMMENTED** [review]
This is great! Just a couple of suggestions, and then we'll need to add it to our API testers but I can do that.

File: RevenueCatUI/View+PresentPaywall.swift (hunk 3 above)
Comment (Lascorbe) [reply RevenueCatUI/View+PresentPaywall.swift:58]: good point! i tend to go DRY (don't repeat yourself) when it's 3 o more

**Comment by Lascorbe** [comment]
should I squash the commits?

**Review by NachoSoto — APPROVED** [review]

**Comment by NachoSoto** [comment]
That's all right, thanks for this! I'm merging in a separate branch so we can run CI on it and update API testers.

File: RevenueCatUI/View+PresentPaywall.swift [hunk 8]
```diff
@@ -37,14 +50,19 @@ extension View {
...
     /// that it will ignore the offering configured in an active experiment.
-    /// - Parameter fonts: An optional `PaywallFontProvider`.
+    /// - Parameter fonts: An optional ``PaywallFontProvider``.
+    /// - Parameter presentationMode: The desired presentation mode of the ``PaywallView`` (``PaywallPresentationMode``).
+    /// Defaults to `.sheet`.
     ///
     /// ### Related Articles
     /// [Documentation](https://rev.cat/paywalls)
+    ///
+    /// - Tag: presentPaywallIfNeeded
```
Comment (NachoSoto) [inline RevenueCatUI/View+PresentPaywall.swift:60]: TIL!
